src/client_event.rs

```diff
@@ -0,0 +1,599 @@
+                    });
+                });
+            });
+        });
+    }
+
+    fn receive(world: &mut World) {
+        world.resource_scope(|world, mut server: Mut<RepliconServer>| {
+            world.resource_scope(|world, registry: Mut<AppTypeRegistry>| {
+                world.resource_scope(|world, event_registry: Mut<ClientEventRegistry>| {
+                    let mut ctx = ServerReceiveCtx {
+                        registry: &registry.read(),
+                    };
+
+                    for event_data in &event_registry.0 {
+                        let client_events = world
+                            .get_resource_mut_by_id(event_data.client_events_id)
+                            .expect("client events shouldn't be removed");
+
+                        // SAFETY: passed pointer was obtained from this event ID.
+                        unsafe {
+                            event_data.receive(&mut ctx, client_events.into_inner(), &mut server)
+                        };
+                    }
+                });
+            });
+        });
+    }
+
+    fn resend_locally(world: &mut World) {
+        world.resource_scope(|world, event_registry: Mut<ClientEventRegistry>| {
+            let world_cell = world.as_unsafe_world_cell();
+            for event_data in &event_registry.0 {
+                // SAFETY: both resources mutably borrowed uniquely.
+                let (events, client_events) = unsafe {
+                    let events = world_cell
+                        .get_resource_mut_by_id(event_data.events_id)
+                        .expect("events shouldn't be removed");
+                    let client_events = world_cell
+                        .get_resource_mut_by_id(event_data.client_events_id)
+                        .expect("client events shouldn't be removed");
+                    (events, client_events)
+                };
+
+                // SAFETY: passed pointers were obtained from this event IDs.
+                unsafe {
+                    event_data.resend_locally(events.into_inner(), client_events.into_inner())
+                };
+            }
+        });
+    }
+
+    fn reset(world: &mut World) {
+        world.resource_scope(|world, event_registry: Mut<ClientEventRegistry>| {
+            for event_data in &event_registry.0 {
+                let events = world
+                    .get_resource_mut_by_id(event_data.events_id)
+                    .expect("events shouldn't be removed");
+
+                // SAFETY: passed pointer was obtained from this event ID.
+                unsafe { event_data.reset(events.into_inner()) };
+            }
+        });
+    }
+}
+
+/// Registered client events.
+#[derive(Resource, Default)]
+struct ClientEventRegistry(Vec<ClientEventData>);
+
+/// Type-erased functions and metadata for a registered client event.
+///
+/// Needed to process all events in a single system and call its functions without knowing the type.
+struct ClientEventData {
```

Comment:
I think it worth to move this and server structs into their own modules.
I will do it after the review in this PR since I marked it as ready for review and don't want to disrupt the process.